crossplane.giantswarm.io/xappdeployment/v1alpha1/appdeployment_types.go

```diff
@@ -0,0 +1,202 @@
+package v1alpha1
+
+import (
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	v2beta1 "github.com/fluxcd/helm-controller/api/v2beta1"
+	"github.com/fluxcd/pkg/apis/meta"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// Repository type metadata.
+var (
+	AppDeploymentKind      = "AppDeployment"
+	AppDeploymentGroupKind = schema.GroupKind{
+		Group: XRDGroup,
+		Kind:  AppDeploymentKind,
+	}.String()
+	AppDeploymentKindAPIVersion   = AppDeploymentKind + "." + GroupVersion.String()
+	AppDeploymentGroupVersionKind = GroupVersion.WithKind(AppDeploymentKind)
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +genclient
+// +genclient:nonNamespaced
+//
+// +kubebuilder:resource:scope=Cluster,categories=crossplane
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:shortName=ad
+// +crossbuilder:generate:xrd:claimNames:kind=AppDeploymentClaim,plural=appdeploymentclaims
+// +crossbuilder:generate:xrd:defaultCompositionRef:name=app-deployment
+type AppDeployment struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   AppDeploymentSpec   `json:"spec"`
+	Status AppDeploymentStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+type AppDeploymentList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []AppDeployment `json:"items"`
+}
+
+// Use the `spec` struct to contain parameters you might not want to share
+// when nesting XRDs - these will usually be parameters that may be defined
+// in a parent.
+
+type AppDeploymentSpec struct {
+	xpv1.ResourceSpec `json:",inline"`
+
+	AppDeploymentParameters `json:",inline"`
+}
+
+type AppDeploymentParameters struct {
+	AppDeploymentConfigMapParameters `json:",inline"`
+
+	// KubernetesProviderConfig is the provider config for the Kubernetes provider.
+	//
+	// +required
+	KubernetesProviderConfig *xpv1.Reference `json:"kubernetesProviderConfig,omitempty"`
+}
+
+type AppDeploymentConfigMapParameters struct {
+	// Name of the configmap.
+	//
+	// +required
+	Name string `json:"name,omitempty"`
+
+	// Namespace of the configmap
+	//
+	// +required
+	Namespace string `json:"namespace,omitempty"`
+
+	// Message to store in the configmap.
+	//
+	// +required
+	Message string `json:"message,omitempty"`
+
+	// Database parameters.
+	//
+	// +optional
+	Database DatabaseParameters `json:"database,omitempty"`
```

Comment:
Is there any reason here to define additional structs for DatabaseParameters? Is it possible instead to rely on the structs defined in https://github.com/giantswarm/crossplane-gs-apis/blob/main/crossplane.giantswarm.io/xdatabase/v1alpha1/rdsprovisioning_types.go?

Reply:
Only engine is used from there, which is a mere string, keeping as is for now.